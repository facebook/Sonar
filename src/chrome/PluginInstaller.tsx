/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {
  FlexColumn,
  styled,
  ManagedTable_immutable,
  Toolbar,
  SearchInput,
  SearchBox,
  Button,
  colors,
  Spacer,
  TableRows_immutable,
  FlexRow,
  Glyph,
  Link,
  Text,
  LoadingIndicator,
  Tooltip,
} from 'flipper';
import React, {useCallback, useState, useMemo, useEffect} from 'react';
import {List} from 'immutable';
import algoliasearch from 'algoliasearch';
import path from 'path';
import fs from 'fs-extra';
import {homedir} from 'os';
import {PluginManager as PM} from 'live-plugin-manager';
import {reportPlatformFailures, reportUsage} from '../utils/metrics';
import restartFlipper from '../utils/restartFlipper';

const PLUGIN_DIR = path.join(homedir(), '.flipper', 'thirdparty');
const ALGOLIA_APPLICATION_ID = 'OFCNCOG2CU';
const ALGOLIA_API_KEY = 'f54e21fa3a2a0160595bb058179bfb1e';
const TAG = 'PluginInstaller';
const PluginManager = new PM({
  ignoredDependencies: ['flipper', 'react', 'react-dom', '@types/*'],
});

export type PluginDefinition = {
  name: string;
  version: string;
  description: string;
};

const EllipsisText = styled(Text)({
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
});

const columnSizes = {
  name: '25%',
  version: '10%',
  description: 'flex',
  install: '15%',
};

const columns = {
  name: {
    value: 'Name',
  },
  version: {
    value: 'Version',
  },
  description: {
    value: 'Description',
  },
  install: {
    value: '',
  },
};

const Container = styled(FlexColumn)({
  height: 300,
  backgroundColor: colors.white,
  borderRadius: 4,
  overflow: 'hidden',
  border: `1px solid ${colors.macOSTitleBarButtonBorder}`,
});

const RestartBar = styled(FlexColumn)({
  backgroundColor: colors.red,
  color: colors.white,
  fontWeight: 500,
  padding: 10,
  cursor: 'pointer',
  textAlign: 'center',
});

type Props = {
  searchIndexFactory: () => algoliasearch.Index;
  getInstalledPlugins: () => Promise<Map<string, PluginDefinition>>;
  autoHeight: boolean;
};

const defaultProps: Props = {
  searchIndexFactory: () => {
    const client = algoliasearch(ALGOLIA_APPLICATION_ID, ALGOLIA_API_KEY);
    return client.initIndex('npm-search');
  },
  getInstalledPlugins: _getInstalledPlugins,
  autoHeight: false,
};

const PluginInstaller = function props(props: Props) {
  const [restartRequired, setRestartRequired] = useState(false);
  const [query, setQuery] = useState('');
  const rows = useNPMSearch(
    setRestartRequired,
    query,
    setQuery,
    props.searchIndexFactory,
    props.getInstalledPlugins,
  );
  const restartApp = useCallback(() => {
    restartFlipper();
  }, []);

  return (
    <Container>
      {restartRequired && (
        <RestartBar onClick={restartApp}>
          To activate this plugin, Flipper needs to restart. Click here to
          restart!
        </RestartBar>
      )}
      <Toolbar>
        <SearchBox>
          <SearchInput
            onChange={e => setQuery(e.target.value)}
            value={query}
            placeholder="Search Flipper plugins..."
          />
        </SearchBox>
      </Toolbar>
      <ManagedTable_immutable
        rowLineHeight={28}
        floating={false}
        multiline={true}
        columnSizes={columnSizes}
        columns={columns}
        highlightableRows={false}
        highlightedRows={new Set()}
        autoHeight={props.autoHeight}
        rows={rows}
      />
    </Container>
  );
};
PluginInstaller.defaultProps = defaultProps;
export default PluginInstaller;

const TableButton = styled(Button)({
  marginTop: 2,
});

const Spinner = styled(LoadingIndicator)({
  marginTop: 6,
});

const AlignedGlyph = styled(Glyph)({
  marginTop: 6,
});

function InstallButton(props: {
  name: string;
  version: string;
  onInstall: () => void;
  installed: boolean;
}) {
  type InstallAction =
    | {kind: 'Install'; error?: string}
    | {kind: 'Waiting'}
    | {kind: 'Remove'; error?: string};

  const catchError = (
    actionKind: 'Install' | 'Remove',
    fn: () => Promise<void>,
  ) => async () => {
    try {
      await fn();
    } catch (err) {
      setAction({kind: actionKind, error: err.toString()});
    }
  };

  const performInstall = useCallback(
    catchError('Install', async () => {
      reportUsage(`${TAG}:install`, undefined, props.name);
      setAction({kind: 'Waiting'});
      await fs.ensureDir(PLUGIN_DIR);
      // create empty watchman config (required by metro's file watcher)
      await fs.writeFile(path.join(PLUGIN_DIR, '.watchmanconfig'), '{}');

      // install the plugin and all it's dependencies into node_modules
      PluginManager.options.pluginsPath = path.join(
        PLUGIN_DIR,
        props.name,
        'node_modules',
      );
      await PluginManager.install(props.name);

      // move the plugin itself out of the node_modules folder
      const pluginDir = path.join(
        PLUGIN_DIR,
        props.name,
        'node_modules',
        props.name,
      );
      const pluginFiles = await fs.readdir(pluginDir);
      await Promise.all(
        pluginFiles.map(f =>
          fs.move(path.join(pluginDir, f), path.join(pluginDir, '..', '..', f)),
        ),
      );

      props.onInstall();
      setAction({kind: 'Remove'});
    }),
    [props.name, props.version],
  );

  const performRemove = useCallback(
    catchError('Remove', async () => {
      reportUsage(`${TAG}:remove`, undefined, props.name);
      setAction({kind: 'Waiting'});
      await fs.remove(path.join(PLUGIN_DIR, props.name));
      props.onInstall();
      setAction({kind: 'Install'});
    }),
    [props.name],
  );

  const [action, setAction] = useState<InstallAction>(
    props.installed ? {kind: 'Remove'} : {kind: 'Install'},
  );

  if (action.kind === 'Waiting') {
    return <Spinner size={16} />;
  }
  if ((action.kind === 'Install' || action.kind === 'Remove') && action.error) {
  }
  const button = (
    <TableButton
      compact
      type={action.kind === 'Install' ? 'primary' : undefined}
      onClick={
        action.kind === 'Install'
          ? () => reportPlatformFailures(performInstall(), `${TAG}:install`)
          : () => reportPlatformFailures(performRemove(), `${TAG}:remove`)
      }>
      {action.kind}
    </TableButton>
  );

  if (action.error) {
    const glyph = (
      <AlignedGlyph color={colors.orange} size={16} name="caution-triangle" />
    );
    return (
      <FlexRow>
        <Tooltip
          options={{position: 'toLeft'}}
          title={`Something went wrong: ${action.error}`}
          children={glyph}
        />
        {button}
      </FlexRow>
    );
  } else {
    return button;
  }
}

function useNPMSearch(
  setRestartRequired: (restart: boolean) => void,
  query: string,
  setQuery: (query: string) => void,
  searchClientFactory: () => algoliasearch.Index,
  getInstalledPlugins: () => Promise<Map<string, PluginDefinition>>,
): TableRows_immutable {
  const index = useMemo(searchClientFactory, []);
  const [installedPlugins, setInstalledPlugins] = useState(
    new Map<string, PluginDefinition>(),
  );

  const getAndSetInstalledPlugins = () =>
    reportPlatformFailures(
      getInstalledPlugins(),
      `${TAG}:getInstalledPlugins`,
    ).then(setInstalledPlugins);

  useEffect(() => {
    reportUsage(`${TAG}:open`);
    getAndSetInstalledPlugins();
  }, []);

  const onInstall = useCallback(async () => {
    getAndSetInstalledPlugins();
    setRestartRequired(true);
  }, []);

  const createRow = useCallback(
    (h: PluginDefinition) => ({
      key: h.name,
      columns: {
        name: {value: <EllipsisText>{h.name}</EllipsisText>},
        version: {
          value: <EllipsisText>{h.version}</EllipsisText>,
          align: 'flex-end' as 'flex-end',
        },
        description: {
          value: (
            <FlexRow grow>
              <EllipsisText>{h.description}</EllipsisText>
              <Spacer />
              <Link href={`https://yarnpkg.com/en/package/${h.name}`}>
                <Glyph color={colors.light20} name="info-circle" size={16} />
              </Link>
            </FlexRow>
          ),
        },
        install: {
          value: (
            <InstallButton
              name={h.name}
              version={h.version}
              onInstall={onInstall}
              installed={installedPlugins.has(h.name)}
            />
          ),
          align: 'center' as 'center',
        },
      },
    }),
    [installedPlugins],
  );

  const [searchResults, setSearchResults] = useState<PluginDefinition[]>([]);

  useEffect(() => {
    (async () => {
      const {hits} = await reportPlatformFailures(
        index.search({
          query,
          filters: 'keywords:flipper-plugin',
          hitsPerPage: 20,
        }),
        `${TAG}:queryIndex`,
      );

      setSearchResults(hits.filter(hit => !installedPlugins.has(hit.name)));
      setQuery(query);
    })();
  }, [query, installedPlugins]);

  const results = Array.from(installedPlugins.values()).concat(searchResults);
  return List(results.map(createRow));
}

async function _getInstalledPlugins(): Promise<Map<string, PluginDefinition>> {
  const pluginDirExists = await fs.pathExists(PLUGIN_DIR);

  if (!pluginDirExists) {
    return new Map();
  }
  const dirs = await fs.readdir(PLUGIN_DIR);
  const plugins = await Promise.all<[string, PluginDefinition]>(
    dirs.map(
      name =>
        new Promise(async (resolve, reject) => {
          if (!(await fs.lstat(path.join(PLUGIN_DIR, name))).isDirectory()) {
            return resolve(undefined);
          }

          const packageJSON = await fs.readFile(
            path.join(PLUGIN_DIR, name, 'package.json'),
          );

          try {
            resolve([name, JSON.parse(packageJSON.toString())]);
          } catch (e) {
            reject(e);
          }
        }),
    ),
  );
  return new Map(plugins.filter(Boolean));
}
