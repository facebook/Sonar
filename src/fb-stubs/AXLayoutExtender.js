/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import {Component} from 'react';
import type {Element, ElementID, ElementSearchResultSet} from 'sonar';

export class AXElementsInspector extends Component<{
  onElementExpanded: (key: ElementID, deep: boolean) => void,
  onElementSelected: (key: ElementID) => void,
  onElementHovered: ?(key: ?ElementID) => void,
  onValueChanged: ?(path: Array<string>, val: any) => void,
  selected: ?ElementID,
  searchResults?: ?ElementSearchResultSet,
  root: ?ElementID,
  elements: {[key: ElementID]: Element},
  useAppSidebar?: boolean,
}> {
  render() {
    return null;
  }
}
