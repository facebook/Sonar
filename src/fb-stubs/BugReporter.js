/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import LogManager from './Logger';

import type {Store} from '../reducers/index.js';

export default class BugReporter {
  constructor(logManager: LogManager, store: Store) {}
  async report(title: string, body: string): Promise<number> {
    return Promise.resolve(-1);
  }
}
