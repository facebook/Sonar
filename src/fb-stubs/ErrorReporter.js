/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 * @format
 */

export function cleanStack(stack: string, loc: ?string) {}
import type ScribeLogger from './ScribeLogger';

export type ObjectError =
  | Error
  | {
      message: string,
      stack?: string,
    };

export default class ErrorReporter {
  constructor(scribeLogger: ScribeLogger) {}
  report(err: ObjectError) {}
}
