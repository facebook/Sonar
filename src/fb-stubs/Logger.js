/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import type {TrackType, Logger} from '../fb-interfaces/Logger';
import type {Store} from '../reducers/index';
import ScribeLogger from './ScribeLogger';

var instance: ?StubLogger = null;

type Args = {
  isHeadless?: boolean,
};

class StubLogger implements Logger {
  constructor(store: Store, args?: Args) {
    this.scribeLogger = new ScribeLogger(this);
  }

  scribeLogger: ScribeLogger;

  track(type: TrackType, event: string, data: ?any, plugin?: string) {}

  trackTimeSince(mark: string, eventName: ?string) {}

  info(data: any, category: string) {}

  warn(data: any, category: string) {}

  error(data: any, category: string) {}

  debug(data: any, category: string) {}
}

export function init(store: Store, args?: Args): Logger {
  if (instance) {
    throw new Error('Attempted to initialize Logger when already initialized');
  }
  instance = new StubLogger(store);
  return instance;
}

export function getInstance(): Logger {
  if (!instance) {
    throw new Error(
      'Requested Logger instance without initializing it. Make sure init() is called at app start',
    );
  }
  return instance;
}
