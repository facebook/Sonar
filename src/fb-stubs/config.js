/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

export default {
  updateServer: 'https://www.facebook.com/sonar/public/latest.json',
  bugReportButtonVisible: false,
};
