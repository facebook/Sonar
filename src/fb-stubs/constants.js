/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 * @format
 */

export const GRAPH_APP_ID = '';
export const GRAPH_CLIENT_TOKEN = '';
export const GRAPH_ACCESS_TOKEN = '';

// this provides elevated access to scribe. we really shouldn't be exposing this.
// need to investigate how to abstract the scribe logging so it's safe.
export const GRAPH_SECRET = '';
export const GRAPH_SECRET_ACCESS_TOKEN = '';

// Provides access to Insights Validation ednpoint on interngraph
export const INSIGHT_INTERN_APP_ID = '';
export const INSIGHT_INTERN_APP_TOKEN = '';
