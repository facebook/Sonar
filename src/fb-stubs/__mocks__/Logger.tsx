/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {Store} from '../../reducers/index';
import {getStringFromErrorLike} from '../../utils/errors';
import {Args, Logger, TrackType} from '../../fb-interfaces/Logger';

export function extractError(
  ...data: Array<any>
): {message: string; error: Error} {
  const message = data.map(getStringFromErrorLike).join(' ');
  const error = data.find(e => e instanceof Error) || new Error(message);
  return {
    message,
    error,
  };
}

export class FBLogger implements Logger {
  constructor(_store?: Store, _args?: Args) {}

  track(_type: TrackType, _event: string, _data?: any, _plugin?: string) {}

  trackTimeSince(_mark: string, _eventName?: string) {}

  info = (..._data: Array<any>) => {};

  warn = (..._data: Array<any>) => {};

  error = (..._data: Array<any>) => {};

  debug = (..._data: Array<any>) => {};

  getLogs(): Array<string> {
    return [];
  }
}

export function init(_store: Store, _args?: Args): Logger {
  return new FBLogger();
}

export function getInstance(): Logger {
  return new FBLogger();
}
