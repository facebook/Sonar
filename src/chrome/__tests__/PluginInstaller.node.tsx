/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import {default as PluginInstaller, PluginDefinition} from '../PluginInstaller';

import React from 'react';
import {render, waitForElement} from '@testing-library/react';
import {init as initLogger} from '../../fb-stubs/Logger';
import configureStore from 'redux-mock-store';

const mockStore = configureStore([])({application: {sessionId: 'mysession'}});
import {Provider} from 'react-redux';

const SEARCH_RESULTS = ({
  hits: [
    {name: 'flipper-plugin-hello', version: '0.1.0', description: 'World?'},
    {name: 'flipper-plugin-world', version: '0.2.0', description: 'Hello?'},
  ],
} as unknown) as algoliasearch.Response<any>;

// *Very* incomplete mock, but that's all we use.
const indexMock: algoliasearch.Index = ({
  search: jest.fn(),
} as unknown) as algoliasearch.Index;

beforeEach(() => {
  indexMock.search = jest.fn(async () => SEARCH_RESULTS);
  initLogger(mockStore as any, {isTest: true});
});

test('load PluginInstaller list', async () => {
  const component = (
    <Provider store={mockStore}>
      <PluginInstaller
        getInstalledPlugins={async () => new Map<string, PluginDefinition>()}
        searchIndexFactory={() => indexMock}
        // Bit ugly to have this as an effectively test-only option, but
        // without, we rely on height information from Electron which we don't
        // have, causing no items to be rendered.
        autoHeight={true}
      />
    </Provider>
  );
  const {container, getByText} = render(component);
  await waitForElement(() => getByText('flipper-plugin-hello'));
  expect((indexMock.search as jest.Mock).mock.calls.length).toBe(2);
  expect(container).toMatchSnapshot();
});
