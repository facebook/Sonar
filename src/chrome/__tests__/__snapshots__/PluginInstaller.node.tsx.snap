// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load PluginInstaller list 1`] = `
<div>
  <div
    class="css-1nm3y77"
  >
    <div
      class="css-4zga7z"
    >
      <div
        class="css-e0frhe"
      >
        <input
          class="css-ww9vf1"
          placeholder="Search Flipper plugins..."
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="css-1nekdsz"
    >
      <div
        class="css-1k4677w"
      >
        <div
          class="css-vd30c4"
        >
          <div
            class="css-1vb6sy6"
            title="name"
            width="25%"
          >
            <div
              class="css-glr1wj"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1db3q1"
              >
                Name
              </div>
            </div>
          </div>
          <div
            class="css-1xchw24"
            title="version"
            width="10%"
          >
            <div
              class="css-glr1wj"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1db3q1"
              >
                Version
              </div>
            </div>
          </div>
          <div
            class="css-173sh01"
            title="description"
            width="flex"
          >
            <div
              class="css-glr1wj"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1db3q1"
              >
                Description
              </div>
            </div>
          </div>
          <div
            class="css-f0p3z5"
            title="install"
            width="15%"
          >
            <div
              class="css-glr1wj"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1db3q1"
              >
                
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-1nekdsz"
      >
        <div
          class="css-1jjayz9"
          data-key="flipper-plugin-hello"
        >
          <div
            class="css-1jes4os"
            title=""
            width="25%"
          >
            <span
              class="css-1pso8q0"
            >
              flipper-plugin-hello
            </span>
          </div>
          <div
            class="css-1kkefm9"
            title=""
            width="10%"
          >
            <span
              class="css-1pso8q0"
            >
              0.1.0
            </span>
          </div>
          <div
            class="css-v10b1x"
            title=""
            width="flex"
          >
            <div
              class="css-9qtipk"
            >
              <span
                class="css-1pso8q0"
              >
                World?
              </span>
              <div
                class="css-12zzrdt"
              />
              <span
                class="css-1b8mb9l"
              >
                <div
                  class="css-1fev88q"
                  color="#bec2c9"
                  size="16"
                  src="https://external.xx.fbcdn.net/assets/?name=info-circle&variant=filled&size=16&set=facebook_icons&density=1x"
                />
              </span>
            </div>
          </div>
          <div
            class="css-yt1ntn"
            title=""
            width="15%"
          >
            <div
              class="css-z1ci4f"
              type="primary"
            >
              Install
            </div>
          </div>
        </div>
        <div
          class="css-1ut44ht"
          data-key="flipper-plugin-world"
        >
          <div
            class="css-1jes4os"
            title=""
            width="25%"
          >
            <span
              class="css-1pso8q0"
            >
              flipper-plugin-world
            </span>
          </div>
          <div
            class="css-1kkefm9"
            title=""
            width="10%"
          >
            <span
              class="css-1pso8q0"
            >
              0.2.0
            </span>
          </div>
          <div
            class="css-v10b1x"
            title=""
            width="flex"
          >
            <div
              class="css-9qtipk"
            >
              <span
                class="css-1pso8q0"
              >
                Hello?
              </span>
              <div
                class="css-12zzrdt"
              />
              <span
                class="css-1b8mb9l"
              >
                <div
                  class="css-1fev88q"
                  color="#bec2c9"
                  size="16"
                  src="https://external.xx.fbcdn.net/assets/?name=info-circle&variant=filled&size=16&set=facebook_icons&density=1x"
                />
              </span>
            </div>
          </div>
          <div
            class="css-yt1ntn"
            title=""
            width="15%"
          >
            <div
              class="css-z1ci4f"
              type="primary"
            >
              Install
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
